### ISSUE TITLE
Resumable training

### ISSUE THREAD

**Issue by roomrys** [talmolab/sleap#1117]
# Problem Description
- In order to avoid running into a local minimum when training, SLEAP always restarts training from scratch.
- Thus, to get to the "same := similar performance" model weights as a previously trained model takes some time.

# Feature Proposal
- Instead of always training models from scratch, let's add an option for resumable training.
- Initialize the new model to have the same weights as the previously trained model.

# Implementation Details
Currently, SLEAP allows user's to "Use Trained Model", but this does not actually train the model any further - the weights from the previous model are just used for inference. We would like to instead initialize a new model with the weights of the trained model and resume training using all the available `LabeledFrame`s. The option to do this should both be available in the GUI and in the CLI.

PR 1: Add functionality to resume training
---

### 1. Add a `resume_training` attribute to the [`ModelConfig` class](https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/nn/config/model.py#L649-L659)
- [x] Initialize `resume_training` to `False`
- [x] Add a docstring describing what `resume_training` affects/does
https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/nn/config/model.py#L649-L659

### 2. Load the weights from the trained model if `Trainer.config.model.resume_training == True`
- [x] Edit the `Trainer._setup_model` function: add `self.keras_model.load_weights(self.config.filename)`
https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/nn/training.py#L721-L746

PR 2: Add CLI option to resume training
--- 
For [the CLI](https://sleap.ai/guides/cli.html), SLEAP uses the [argparse](https://docs.python.org/3/library/argparse.html) library to register and parse all known arguments.
### 1. Add a new optional argument `--resume` to `sleap-train`
- [x] Register the new argument here with `parser.add_argument`, the default should be `False` or `0`:
https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/nn/training.py#L1788-L1886
- [x] Copy the `help` string that you wrote in the previous step to the CLI documentation:
https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/docs/guides/cli.md?plain=1#L50-L79
- [x] Set `job_config.model.resume_training` to `args.resume`
https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/nn/training.py#L1899-L1905

PR 3: Expose CLI option to GUI (Training Pipeline)
---

### 1. Add a "Resume Training" checkbox item in [`TrainingEditorWidget`](https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/learning/dialog.py#L840) 

The Training Pipeline option "Use Trained Model" is what we want to model this PR off of (use as a template). Most of the options seen in the Training/Inference Pipelines are listed in the [sleap/configs/pipeline_form.yaml](https://github.com/talmolab/sleap/blob/develop/sleap/config/pipeline_form.yaml) which are then used by the [`YamlFormWidget`](https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/dialogs/formbuilder.py#L40) class to create a `QtWidget` (GUI). However, since the "Use Trained Model" option is dynamically displayed iff a trained model is chosen, the code to add the "Use Trained Model" is actually spread throughout the [`TrainingEditorWidget`](https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/learning/dialog.py#L840) class. (This could be refactored to be listed in a yaml like all the other options, but we'll save that for another PR).

The keyword to search for is `_use_trained_model` - this is the `TrainingEditorWidget` attribute which contains the "Use Trained Model" widget (which we are using as a template).
- [x] Initialize a `_resume_training` attribute to `None`
https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/learning/dialog.py#L859-L875
- [x] Add [`QCheckBox`](https://doc.qt.io/qt-6/qcheckbox.html) for resuming training for trained model from selected config. Connect the checkbox to `_update_use_trained` as well.
https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/learning/dialog.py#L950-L960
- [x] Modify `_update_use_training`: reset the form to match the config when `_resume_training` is checked (but this should not lock the fields like `_use_trained_model` does - unless `_use_trained_model` is checked as well).
https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/learning/dialog.py#L1035-L1047
- [x] Update the visibility of the `_resume_training` checkbox
https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/learning/dialog.py#L989-L995
- [x] Add an attribute `resume_training` which uses the [`@property` decorator](https://www.programiz.com/python-programming/property) and returns whether `_resume_training` (is visible and checked (see: [QCheckBox.checkState()](https://doc.qt.io/qt-6/qcheckbox.html#checkState) and [enum Qt::CheckState](https://doc.qt.io/qt-6/qt.html#CheckState-enum)).
https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/learning/dialog.py#L1074-L1081
- [x] Set the `trained_config_info.model.resume_training` attribute to `True` if `resume_training` is `True`
https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/learning/dialog.py#L1083-L1099
- [x] Modify the above code s.t. `dont_retrain` is `True` iff `self.use_trained and (not self.resume_training)`

---

- Add checkbox to trainer configuration GUI to “Load existing checkpoint if available”
- Add field to TrainingJobConfig pointing to base checkpoint
- In Trainer subclasses, load weights (or just load the model architecture) from checkpoint when building model

**Comment by jimzers** [comment]
#1130

**Comment by roomrys** [comment]
This feature is now available in the (pre) release 1.3.0a0, to install, first [uninstall](https://sleap.ai/installation.html#upgrading-and-uninstalling) and then:
**`conda` (Windows/Linux/GPU)**:
```
conda create -y -n sleap -c sleap -c sleap/label/dev -c nvidia -c conda-forge sleap=1.3.0a0
```
**`pip` (any OS except Apple Silicon)**:
```
pip install sleap==1.3.0a0
```

**Warning:** This is a pre-release! Expect bugs and strange behavior when testing.
